crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -488,42 +453,31 @@ where
             };
 
             // Find plot in which there is a place for new piece to be stored
-            let mut sorted_caches = caches.iter_mut().enumerate().collect::<Vec<_>>();
-            // Sort piece caches by number of stored pieces to fill those that are less
-            // populated first
-            sorted_caches.sort_by_key(|(_, cache)| cache.stored_pieces.len());
-            if !stream::iter(sorted_caches)
-                .any(|(cache_index, cache)| async move {
-                    let Some(offset) = cache.free_offsets.pop_front() else {
-                        return false;
-                    };
+            let Some(offset) = caches.free_offsets.pop_front() else {
+                error!(
+                    %piece_index,
+                    "Failed to store piece in cache, there was no space"
+                );
+                break;
+            };
 
-                    if let Err(error) = cache.backend.write_piece(offset, *piece_index, piece).await
-                    {
-                        error!(
-                            %error,
-                            %cache_index,
-                            %piece_index,
-                            %offset,
-                            "Failed to write piece into cache"
-                        );
-                        return false;
-                    }
-                    if let Some(metrics) = &self.metrics {
-                        metrics.piece_cache_capacity_used.inc();
-                    }
-                    cache
-                        .stored_pieces
-                        .insert(RecordKey::from(piece_index.to_multihash()), offset);
-                    true
-                })
-                .await
+            let cache_index = usize::from(offset.cache_index);
+            let piece_offset = offset.piece_offset;
+            if let Some(backend) = caches.backends.get(cache_index)
+                && let Err(error) = backend.write_piece(piece_offset, *piece_index, piece).await
             {
                 error!(
+                    %error,
+                    %cache_index,
                     %piece_index,
-                    "Failed to store piece in cache, there was no space"
+                    %piece_offset,
+                    "Failed to write piece into cache"
                 );
+                break;
```

Comment:
The fact that one backend has errors shouldn't prevent pieces from being stored in other backends. See how previous code was iterating over all caches to try and find one that is operational. We still need that loop here, but the issue is that we can't iterate over different backends anymore, so it will be less performant either way (will likely need to cache problematic backend indices to avoid hitting it over and over again repeatedly).

Reply:
You're right. It could be a disaster here.